core/domain/classifier_services.py

```diff
@@ -0,0 +1,322 @@
+        }
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            for c in xrange(len(doc)):
+                w = doc[c]
+                l = self._l_dc[d][c]
+
+                self._update_counting_matrices(d, w, l, -1)
+
+                # Gibbs update of labels
+                coeff_a = 1 / (
+                    self._c_dl[d].sum() + self._label_count * self._alpha)
+                coeff_b = 1 / (
+                    self._c_lw.sum(axis=1) + self._word_count * self._beta)
+                prob_l = (
+                    labels *
+                    coeff_a * (self._c_dl[d] + self._alpha) *
+                    coeff_b * (self._c_lw[:, w] + self._beta))
+                new_l = numpy.random.multinomial(
+                    1, prob_l / prob_l.sum()
+                    ).argmax()
+
+                statez['computes'] += 1
+                if l != new_l:
+                    statez['updates'] += 1
+
+                self._l_dc[d][c] = new_l
+                self._update_counting_matrices(d, w, new_l, 1)
+
+        return statez
+
+    def _get_probabilities(self, d):
+        """Returns the probability of a document having a label."""
+        probs_i = self._c_dl[d] + (self._b_dl[d] * self._alpha)
+        probs_i = probs_i / probs_i.sum(axis=0)[numpy.newaxis]
+        return probs_i
+
+    def _get_probabilities_with_label(self, d):
+        """Includes human readable labels with document probabilities."""
+        probs_i = self._get_probabilities(d)
+        probs_l = {}
+        for label, l in self._label_to_id.iteritems():
+            probs_l[label] = probs_i[l]
+        return probs_l
+
+    def _predict(self, d, threshold=0.5):
+        """Calculates prediction data for a doc"""
+        probs_l = self._get_probabilities_with_label(d)
+        default_prob = probs_l['_default']
+        for l in sorted(probs_l, key=probs_l.get, reverse=True):
+            prob_l = probs_l[l]
+            if l != '_default' and prob_l / (1 - default_prob) > threshold:
+                return l, prob_l / (1 - default_prob), probs_l
+        return '_default', default_prob / (1 - default_prob), probs_l
+
+    def _validate_label(self, label):
+        """Checks that labels don't begin with underscore."""
+        if label[0] == '_':
+            raise Exception('Label %s cannot begin with underscore.' % label)
+
+    def _parse_examples(self, examples):
+        """Splits examples into docs (split on spaces) and labels."""
+        docs = []
+        labels = []
+        for example in examples:
+            doc = example[0].split()
+            if len(doc) > 0:
+                docs.append(doc)
+                map(self._validate_label, example[1])
+                labels.append(example[1])
+        return docs, labels
+
+    def _train_docs(self, iterations=25, doc_ids=None):
+        """Trains given docs (default all) for a number of iterations."""
+        doc_ids = self._get_doc_ids(doc_ids)
```

Comment:
Originally had that. The only downside is if we decide to add more functionality or change how it works, we need to remember to alter this in three different places. I think this will still avoid headaches in the future. LMK if you disagree -or would a different name help? init_doc_ids_if_none?